### ISSUE TITLE
Finding the volume of each element

### ISSUE THREAD

**Issue by ctjacobs** [firedrakeproject/firedrake#183]
I'm trying to get the volume (or area) of each element in order to implement an LES turbulence model. In my DOLFIN implementation I was doing something like:

```
mesh = UnitSquareMesh(10, 10)
fs = FunctionSpace(mesh, "CG", 1)
v = fs.cell().volume
```

Is there a way to do this in Firedrake? Running the above code gives:

```
christian@elevate ~ $ python cell_volume.py 
Info : Increasing process stack size (8192 kB < 16 MB)
Traceback (most recent call last):
  File "cell_volume.py", line 4, in <module>
    v = fs.cell().volume
AttributeError: 'FunctionSpace' object has no attribute 'cell'
```

I've also tried using `CellVolume(mesh)` and `Cell(mesh, 0).volume()`, but I get the same error message for both:

```
christian@elevate ~ $ python cell_volume.py 
Info : Increasing process stack size (8192 kB < 16 MB)
Invalid cellname <firedrake.mesh.UnitSquareMesh object at 0x35b1550>.
Traceback (most recent call last):
  File "cell_volume.py", line 9, in <module>
    c = Cell(mesh, 0).volume()
  File "/usr/local/lib/python2.7/dist-packages/ufl/geometry.py", line 392, in __init__
    ufl_assert(cellname in ufl_cellnames, "Invalid cellname %s." % (cellname,))
  File "/usr/local/lib/python2.7/dist-packages/ufl/assertions.py", line 37, in ufl_assert
    if not condition: error(*message)
  File "/usr/local/lib/python2.7/dist-packages/ufl/log.py", line 154, in error
    raise self._exception_type(self._format_raw(*message))
ufl.log.UFLException: Invalid cellname <firedrake.mesh.UnitSquareMesh object at 0x35b1550>.
```

**Comment by wence-** [comment]
For simplex meshes, this is handled by upcoming symbolic geometry.  For non-simplex meshes the form compiler just needs to learn to emit the correct code.  How do I compute the volume of a quadratic quad?

**Comment by dorugeber** [comment]
Integrate detJ over the reference cell.

**Comment by wence-** [comment]
Oh yeah, duh, that's what UFL already does in the simplex case, but produces a warning for non-affine cells and doesn't do the transformation:

```python
    def cell_volume(self, o):
        if self._preserve_types[o._ufl_typecode_]:
            return o

        domain = o.ufl_domain()
        if not domain.is_piecewise_linear_simplex_domain():
            # Don't lower for non-affine cells, instead leave it to form compiler
            warning("Only know how to compute the cell volume of an affine cell.")
            return o

        r = self.jacobian_determinant(JacobianDeterminant(domain))
        r0 = ReferenceCellVolume(domain)
        return abs(r * r0)
```

**Comment by dorugeber** [comment]
Yeah, it's non-trivial since you need to perform another integral out-of-band.

**Comment by wence-** [comment]
OK,  I see, we just need to remember to emit an integration loop in tsfc.

**Comment by wence-** [comment]
c.f. tsfc#1

**Comment by wence-** [comment]
Fixed for affine elements in #691. Track the non affine at firedrakeproject/tsfc#1
